File: services/leboncoin-bridge/app/mapper.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Protocol

from app.models import LeboncoinAttribute, LeboncoinListing, LeboncoinLocation


Scalar = str | int | float | bool | None

INVALID_VEHICLE_IDENTITIES = {"", "leboncoin", "lbc", "unknown", "inconnu"}


class AttributeLike(Protocol):
    key: str
    key_label: str | None
    value: Scalar
    value_label: str | None
    values: Sequence[str] | None
    values_label: Sequence[str] | None


class LocationLike(Protocol):
    city: str | None
    city_label: str | None
    zipcode: str | None
    department_name: str | None
    region_name: str | None


class AdLike(Protocol):
    id: int | str
    subject: str
    body: str | None
    brand: str | None
    url: str
    price: float | int | None
    images: Sequence[str]
    attributes: Mapping[str, AttributeLike]
    location: LocationLike | None
    first_publication_date: str | None
    favorites: int | None


def _normalize_attribute_name(value: str | None) -> str:
    if value is None:
        return ""
    return "".join(character.lower() for character in value if character.isalnum())


def get_vehicle_attribute(ad: AdLike, *names: str) -> str | None:
    normalized_names = {_normalize_attribute_name(name) for name in names}
    for attribute in ad.attributes.values():
        if any(
            _normalize_attribute_name(candidate) in normalized_names
            for candidate in (attribute.key, attribute.key_label)
        ):
            value = attribute.value_label or attribute.value
            if value is not None and str(value).strip():
                return str(value).strip()
    return None


def credible_vehicle_identity(value: str | None) -> str | None:
    if value is None:
        return None
    cleaned = value.strip()
    if cleaned.lower() in INVALID_VEHICLE_IDENTITIES:
        return None
    return cleaned


def map_listing(ad: AdLike) -> LeboncoinListing:
    attributes = {
        name: LeboncoinAttribute(
            key=attribute.key,
            key_label=attribute.key_label,
            value=attribute.value,
            value_label=attribute.value_label,
            values=list(attribute.values or []),
            values_label=list(attribute.values_label or []),
        )
        for name, attribute in ad.attributes.items()
    }
    location = None
    if ad.location is not None:
        location = LeboncoinLocation(
            city=ad.location.city,
            city_label=ad.location.city_label,
            zipcode=ad.location.zipcode,
            department_name=ad.location.department_name,
            region_name=ad.location.region_name,
        )

    vehicle_brand = get_vehicle_attribute(ad, "u_car_brand", "brand", "marque")
    vehicle_model = get_vehicle_attribute(ad, "u_car_model", "model", "modèle", "modele")

    return LeboncoinListing(
        id=str(ad.id),
        subject=ad.subject,
        body=ad.body,
        brand=credible_vehicle_identity(vehicle_brand) or credible_vehicle_identity(ad.brand),
        model=credible_vehicle_identity(vehicle_model),
        url=ad.url,
        price=float(ad.price) if ad.price is not None else None,
        images=list(ad.images),
        attributes=attributes,
        location=location,
        owner_type="unknown",
        first_publication_date=ad.first_publication_date,
        favorite_count=ad.favorites,
    )

File: src/features/market/providers/leboncoin/mapper.ts
import type { MarketListing, MarketSellerType } from "../../types"
import type { LeboncoinAttribute, LeboncoinListing } from "./types"

const PROVIDER_ID = "leboncoin"

function getAttribute(
  listing: LeboncoinListing,
  ...keys: string[]
): LeboncoinAttribute | undefined {
  const normalizedKeys = new Set(keys.map(normalizeKey))
  return Object.values(listing.attributes).find((attribute) =>
    [attribute.key, attribute.keyLabel].some((value) => value && normalizedKeys.has(normalizeKey(value)))
  )
}

function normalizeKey(value: string) {
  return value.normalize("NFD").replace(/[\u0300-\u036f]/g, "").toLowerCase().replace(/[^a-z0-9]/g, "")
}

function getAttributeText(listing: LeboncoinListing, ...keys: string[]) {
  const attribute = getAttribute(listing, ...keys)
  return attribute?.valueLabel ?? attribute?.value ?? null
}

function getAttributeNumber(listing: LeboncoinListing, ...keys: string[]) {
  const rawValue = getAttributeText(listing, ...keys)
  if (!rawValue) return null

  const normalizedValue = rawValue.replace(/[^0-9.,-]/g, "").replace(",", ".")
  const value = Number(normalizedValue)
  return Number.isFinite(value) ? value : null
}

const INVALID_IDENTITIES = new Set(["", "leboncoin", "lbc", "unknown", "inconnu"])

function credibleIdentity(...values: Array<string | null | undefined>) {
  return values.find((value) => {
    if (!value) return false
    const normalized = value.trim().toLowerCase()
    return !INVALID_IDENTITIES.has(normalized) && /[a-zÀ-ÿ0-9]/i.test(normalized)
  })?.trim() ?? ""
}

function mapSellerType(ownerType: LeboncoinListing["ownerType"]): MarketSellerType {
  if (ownerType === "professional") return "PROFESSIONAL"
  if (ownerType === "private") return "PRIVATE"
  return "UNKNOWN"
}

export function mapLeboncoinListing(listing: LeboncoinListing): MarketListing {
  return {
    providerId: PROVIDER_ID,
    externalId: listing.id,
    url: listing.url,
    title: listing.subject,
    brand: credibleIdentity(
      getAttributeText(listing, "u_car_brand", "brand", "marque"),
      listing.brand
    ),
    model: credibleIdentity(
      getAttributeText(listing, "u_car_model", "model", "modele"),
      listing.model
    ),
    trim: getAttributeText(listing, "vehicle_version", "version", "trim", "finition constructeur", "version constructeur"),
    year:
      getAttributeNumber(listing, "regdate", "registration_year", "year", "annee modele") ??
      null,
    mileage: getAttributeNumber(listing, "mileage", "kilometrage"),
    fuel: getAttributeText(listing, "fuel", "energie", "carburant"),
    gearbox: getAttributeText(listing, "gearbox", "boite de vitesses"),
    powerDin: getAttributeNumber(
      listing,
      "horse_power_din",
      "power_din",
      "horsepower"
    ),
    price: listing.price ?? 0,
    currency: "EUR",
    location:
      listing.location?.cityLabel ??
      listing.location?.city ??
      listing.location?.departmentName ??
      null,
    sellerType: mapSellerType(listing.ownerType),
    publishedAt: listing.firstPublicationDate,
    imageUrls: [...listing.images],
    favoriteCount: listing.favoriteCount,
  }
}
